frameworks/Python/klein/app.py

```diff
@@ -67,57 +64,72 @@ def getQueryNum(queryString):
+        finally:
+            db_session.close()
+    return wrapper
+
+
 @app.route("/plaintext")
 def plaintext(request):
-	request.setHeader("Content-Type", "text/plain; charset=UTF-8")
-	return "Hello, World!"
+    request.setHeader("Content-Type", "text/plain; charset=UTF-8")
+    return "Hello, World!"
 
 @app.route("/json")
 def jsonHandler(request):
-	request.setHeader("Content-Type", "application/json; charset=UTF-8")
-	return json.dumps({"message": "Hello, World!"})
+    request.setHeader("Content-Type", "application/json; charset=UTF-8")
+    return json.dumps({"message": "Hello, World!"})
 
 @app.route("/db")
+@close_session
 def db(request):
-	request.setHeader("Content-Type", "application/json; charset=UTF-8")	
-	wid = randint(1, 10000)
-	world = db_session.query(World).get(wid).serialize() 
-	return json.dumps(world)
+    request.setHeader("Content-Type", "application/json; charset=UTF-8")
+    wid = randint(1, 10000)
+    world = db_session.query(World).get(wid).serialize()
+    return json.dumps(world)
 
 @app.route("/queries")
+@close_session
 def queries(request):
-	request.setHeader("Content-Type", "application/json; charset=UTF-8")	
-	num_queries = getQueryNum(request.args.get("queries")[0])
-	rp = partial(randint, 1, 10000)
-	get = db_session.query(World).get
-	worlds = [get(rp()).serialize() for _ in xrange(num_queries)]
-	return json.dumps(worlds)
+    request.setHeader("Content-Type", "application/json; charset=UTF-8")
+    num_queries = getQueryNum(request.args.get("queries")[0])
+    rp = partial(randint, 1, 10000)
+    get = db_session.query(World).get
+    worlds = [get(rp()).serialize() for _ in xrange(num_queries)]
+    return json.dumps(worlds)
 
 @app.route("/updates")
+@close_session
 def updates(request):
-	request.setHeader("Content-Type", "application/json; charset=UTF-8")
-	num_queries = getQueryNum(request.args.get("queries")[0])
-	worlds = []
-	rp = partial(randint, 1, 10000)
-	ids = [rp() for _ in xrange(num_queries)]
-	ids.sort()
-	for id in ids:
-		world = db_session.query(World).get(id)
-		world.randomNumber = rp()
-		worlds.append(world.serialize())
-	db_session.commit()
-	return json.dumps(worlds)
+    request.setHeader("Content-Type", "application/json; charset=UTF-8")
+    num_queries = getQueryNum(request.args.get("queries")[0])
+    worlds = []
+    rp = partial(randint, 1, 10000)
+    ids = [rp() for _ in xrange(num_queries)]
+    ids.sort()
+    for id in ids:
+        world = db_session.query(World).get(id)
+        world.randomNumber = rp()
+        worlds.append(world.serialize())
+    db_session.commit()
+    return json.dumps(worlds)
 
 @app.route("/fortune")
+@close_session
 def fortune(request):
-	request.setHeader("Content-Type", "text/html; charset=UTF-8")
-	fortunes = db_session.query(Fortune).all()
-	fortunes.append(Fortune(id=0, message="Additional fortune added at request time."))
-	fortunes.sort(key=attrgetter("message"))
-	for f in fortunes:
-		f.message = bleach.clean(f.message)
```

Comment:
Yes. SQLAlchemy is "Unit of Work" ORM.
When you modify object which is managed by session and commit the session, all modified objects are updated.

There are no `db_session.commit()` in this request. But since this request doesn't remove session, the session alive
after this request.
When calling "/update" test, modified objects are actually written to DB.

Transaction over HTTP request is nightmare.  So, closing session for each request is required.